### ISSUE TITLE
Beginner Issues

### ISSUE THREAD

**Issue by e1528532** [ElektraInitiative/libelektra#1084]
These are some issues, unclear things and general thoughts i had while starting with elektra for the first time:

# Installation

I am using Mac OS Sierra. Since there is no package for OSX i ended up compiling it myself. After having installed cmake using brew compiling worked as described in the tutorial. For me it was not an issue, but maybe its worth mentioning how to use brew and install cmake on mac?

Next there is a slight inconsistency, on the main page of libelektra.org it states that you issue the following commands to build:

mkdir -p build
cd build
cmake ..
make

On the Compilation Tutorial it suggests mkdir build && cd build && ccmake .. instead. I think that the latter should be also on the main page, for me the ui provided by ccmake made it much easier to get an overview of the available options during compilation and it also was easier to understand how to include/exclude certain bindings or plugins. And i think the -p option on mkdir serves no real purpose in that case and should be removed, i was a bit surprised by the -p option first as i didn't knew its purpose and thought it does something very special.

# First Use

So then after having installed libelektra which worked without any major issues i wanted to try a few simple things to see how it works. The first thing i tried is simply adding some key with kdb set "/test/key1" "value1". However, this command failed and i had no idea why. After checking the tutorial a bit i finally noticed that i HAVE to use one of the available namespaces like /user/ or /system/ at the beginning of the key. The error message just said its not a valid keyname, maybe we should add information about the namespaces on that warning? Initially i really had no idea why its not a valid keyname and thought that maybe there is some issue with my system. But i think this console video of the usage really helps to grasp the first concepts then!

# Advanced Uses

So after trying some basic key setting, mounting and exporting which was quite clear after watching the video i wanted to try out the java bindings, i have created another issue #1083 about my experience with the jni/jna bindings.

A last thing i wondered is how to install the command kdb-full. It is mentioned for example in the documentation of https://github.com/ElektraInitiative/libelektra/tree/master/src/plugins/jni but i havent found out yet how to get this command.

# TL;DR and TODOs

- [x] Use the ccmake variant of compiling on the main page of libelektra.org since i think its easier to grasp the compilation options with that?
- [x] Make it more clear that you can only use specified namespaces in the key name
- [x] Improve the invalid key message if the namespace is wrong
- [x] Explain somewhere how to install kdb-full? I still havent figured out

**Comment by markus2330** [comment]
> brew and install cmake on mac?

Yes, would be good to have a short tutorial covering such issues. It looks like you found a lot of improvements in the "getting started" material, thank you! I will go into the details later. I would like to prefer to review concrete suggestions in PRs, though.

**Comment by markus2330** [comment]
> Since there is no package for OSX i ended up compiling it myself.

See #531 if you are interested.

> `-p` option

It has the advantage to not fail if the directory exists. But you are right, at this point it only confuses.

> kdb set "/test/key1" "value1"

This should work. Do you know the exact command which did not work? (in the example above a default namespace should be assumed) A known problem due to this assumption, however, is #401.

> The error message just said its not a valid keyname

You are absolutely right, we should point to an explanation of key-names: elektra-key-names(7). Error messages could be much user friendly, a first start happened in 42d786f63078fe6294ee02fb18f6addc369224a8
Can you create a PR?

> but i havent found out yet how to get this command.

I do not think you should (have) to use it, but you get it when compiling with cmake flag -DBUILD_FULL=ON.

**Comment by e1528532** [comment]
So i have started working on some documentation improvements in my branch https://github.com/e1528532/libelektra/tree/documentation-improvements .

>See #531 if you are interested.

Will keep it in mind, seems like a nice way to improve the experience on a mac!

> This should work. Do you know the exact command which did not work?

My bad. What does not work is something like kdb set "test/key1" "value1", with the preceding slash it works fine.

> messages could be much user friendly, a first start happened in 42d786f Can you create a PR?

Like said, started working on that, but still needs some further effort before its worth being made a PR. I noticed that in the manual pages (the ones in /doc/help ) the naming of the key name parameter is inconsistent. In some pages its called path (like kdb-test.md or kdb-sget.md), in others mountpoint (like kdb-spec-mount.md), <new path> in kdb-remount.md and in a lot of them its key-name (like kdb-set.md or kdb-get.md)

What do you think makes the most sense, using key-name (and derivates like new key-name or previous key-name) all the time instead of path as elektra basically deals with keys? Or the other way around, using path all the time so its easy to see that elektra acts as a sort of a virtual file system? Because of the latter, i'd be keeping the /mountpoint in the help, but use key-name instead of path (or maybe something like key-path?)

>I do not think you should (have) to use it, but you get it when compiling with cmake flag -DBUILD_FULL=ON.

right, now i have it too. It was referenced in the jni readme, i will probably mention that when improving the jni documentation.

**Comment by markus2330** [comment]
> the naming of the key name parameter is inconsistent

Thanks, improving consistency is always important. In the case of "mountpoint" I think its okay: it is a keyname representing something else (a mountpoint).

> key name vs. key-name vs. keyname vs. key

See #920 

> key-name vs. path

@krit0n made in doc/tutorials/namespaces.md a difference between key (name) and path. @e1528532 What do you think about this explanation (and the difference)?

**Comment by e1528532** [comment]
Done in #1240 .
